### PR TITLE
Support high contrast colors

### PR DESCRIPTION
**Pull request by alexey1312**
Solution for this [issue](https://github.com/RedMadRobot/figma-export/issues/151)

### CODE REVIEW

File: Sources/FigmaExport/Loaders/ColorsLoader.swift [hunk 1]
```diff
@@ -13,34 +13,60 @@ final class ColorsLoader {
...
+                                                       darkHC: [Color]?) {
         let colors = try loadColors(fileId: figmaParams.lightFileId)
         let darkSuffix = colorParams?.darkModeSuffix ?? "_dark"
+        let lightHCSuffix = colorParams?.lightHCModeSuffix ?? "_lightHC"
+        let darkHCSuffix = colorParams?.darkHCModeSuffix ?? "_darkHC"
+
         let lightColors = colors
             .filter { !$0.name.hasSuffix(darkSuffix) }
+            .filter { !$0.name.hasSuffix(lightHCSuffix) }
+            .filter { !$0.name.hasSuffix(darkHCSuffix) }
```
Comment (subdan) [inline Sources/FigmaExport/Loaders/ColorsLoader.swift:47]: Combine three filter operations in to the one to avoid iterating over array three times.

File: Sources/FigmaExportCore/Processor/AssetsProcessor.swift [hunk 2]
```diff
@@ -268,31 +459,31 @@ public extension AssetsProcessable {
...
+                        dark: darkAsset.map { processedAssetName($0) }
+                    )
+                }
+        }
 
     /// Runs the name replacement and name validation regexps, and name styles, if they are defined
     /// - Returns:
     ///   - `AssetType` with a processed name
-    private func processedAssetName(_ asset: AssetType) -> AssetType {
+    func processedAssetName(_ asset: AssetType) -> AssetType {
```
Comment (subdan) [inline Sources/FigmaExportCore/Processor/AssetsProcessor.swift:486]: Why did you remove `private` keyword here an on 501 line?

File: Sources/FigmaExportCore/Processor/AssetsProcessor.swift [hunk 3]
```diff
@@ -200,20 +391,20 @@ public extension AssetsProcessable {
...
         dark.forEach { asset in
             switch darkSet.insert(asset) {
-            case (true, _):
-                break // ok
-            case (false, let oldMember): // already exists
-                errors.all.append(AssetsValidatorError.foundDuplicate(assetName: oldMember.name))
+                case (true, _):
+                    break // ok
+                case (false, let oldMember): // already exists
+                    errors.all.append(AssetsValidatorError.foundDuplicate(assetName: oldMember.name))
```
Comment (subdan) [inline Sources/FigmaExportCore/Processor/AssetsProcessor.swift:407]: Please revert this change.

File: Sources/FigmaExportCore/Processor/AssetsProcessor.swift [hunk 4]
```diff
@@ -179,7 +370,7 @@ public extension AssetsProcessable {
         return .success(pairs)
     }
 
-    private func validateAndMakePairs(light: [AssetType], dark: [AssetType]) -> ProcessingPairResult {
+    func validateAndMakePairs(light: [AssetType], dark: [AssetType]) -> ProcessingPairResult {
```
Comment (subdan) [inline Sources/FigmaExportCore/Processor/AssetsProcessor.swift:373]: Why did you remove `private` keyword here?

**Review by subdan — CHANGES_REQUESTED** [review]
- Add information about supporting high contrast colors in the [Design requirements](https://github.com/RedMadRobot/figma-export#design-requirements) and [Features](https://github.com/RedMadRobot/figma-export#features) section.
- Add unit tests

File: Sources/FigmaExportCore/Processor/AssetsProcessor.swift [hunk 5]
```diff
@@ -88,6 +88,197 @@ public struct ImagesProcessor: AssetsProcessable {
     }
 }
 
+public extension ColorsProcessor {
+    func process(light: [AssetType], dark: [AssetType]?, lightHC: [AssetType]?, darkHC: [AssetType]?) -> ProcessingPairResult {
```
Comment (subdan) [inline Sources/FigmaExportCore/Processor/AssetsProcessor.swift:92]: Why haven't you updated the current method `func process(light: [AssetType], dark: [AssetType]?) -> ProcessingPairResult {`.

File: Sources/FigmaExportCore/Processor/AssetsProcessor.swift [hunk 6]
```diff
@@ -88,6 +88,197 @@ public struct ImagesProcessor: AssetsProcessable {
...
+    }
+
+    private func checkSubtracting(firstAsset: Set<AssetType>, secondAsset: Set<AssetType>, errors: inout ErrorGroup) {
+        let elements = secondAsset.subtracting(firstAsset)
+        if !elements.isEmpty {
+            errors.all.append(AssetsValidatorError.darkAssetsNotFoundInLightPalette(assets: elements.map { $0.name }))
+        }
+    }
+
+    private func foundDuplicate(assets: [AssetType], errors: inout ErrorGroup) -> Set<AssetType> {
```
Comment (subdan) [inline Sources/FigmaExportCore/Processor/AssetsProcessor.swift:200]: Code in this method duplicated in the
`func validateAndMakePairs(light: [AssetType]) -> ProcessingPairResult {`
and
`func validateAndMakePairs(light: [AssetType], dark: [AssetType]) -> ProcessingPairResult {`
method. Could you please remove code duplication?

File: Sources/XcodeExport/XcodeColorExporter.swift [hunk 7]
```diff
@@ -155,15 +156,37 @@ final public class XcodeColorExporter: XcodeExporterBase {
...
+                        color: XcodeAssetContents.ColorInfo(
+                            components: lightHCColor.toHexComponents()
+                        )
+                    )
+                )
+            }
+            if let darkHCColor = colorPair.darkHC {
+                colors.append(
+                    XcodeAssetContents.ColorData(
+                        appearances: [XcodeAssetContents.Appearance(appearance: "contrast", value: "high")],
```
Comment (subdan) [inline Sources/XcodeExport/XcodeColorExporter.swift:186]: You forget to add dark mode appearance here. I noticed it when try to export dark mode high contrast color.

image

File: Sources/FigmaExportCore/Processor/AssetsValidatorError.swift [hunk 8]
```diff
@@ -14,6 +16,10 @@ enum AssetsValidatorError: LocalizedError {
             error = "Bad asset name «\(name)»"
         case .countMismatch(let light, let dark):
             error = "The number of assets doesn’t match. Light theme contains \(light), and dark \(dark)."
+        case .countMismatchLightHighContrastColors(let light, let lightHC):
+            error = "The number of assets doesn’t match. Light color palette contains theme contains \(light), and light high contrast color palette \(lightHC)."
```
Comment (subdan) [inline Sources/FigmaExportCore/Processor/AssetsValidatorError.swift:20]:
```suggestion
            error = "The number of assets doesn’t match. Light color palette contains \(light), and light high contrast color palette \(lightHC)."
```

File: Sources/FigmaExportCore/Processor/AssetsValidatorError.swift [hunk 9]
```diff
@@ -14,6 +16,10 @@ enum AssetsValidatorError: LocalizedError {
...
+        case .countMismatchDarkHighContrastColors(let dark, let darkHC):
+            error = "The number of assets doesn’t match. Dark color palette contains theme contains \(dark), and dark high contrast color palette \(darkHC)."
```
Comment (subdan) [inline Sources/FigmaExportCore/Processor/AssetsValidatorError.swift:22]:
```suggestion
            error = "The number of assets doesn’t match. Dark color palette contains \(dark), and dark high contrast color palette \(darkHC)."
```

File: Sources/XcodeExport/XcodeColorExporter.swift (hunk 7 above)
Comment (subdan) [reply Sources/XcodeExport/XcodeColorExporter.swift:186]: Also, try to avoid code duplication in this method.

File: Sources/FigmaExportCore/Helpers/Zip4Sequence.swift [hunk 10]
```diff
@@ -0,0 +1,203 @@
+
+/// Creates a sequence of tuples built out of 4 underlying sequences.
+///
+/// In the `Zip4Sequence` instance returned by this function, the elements of
+/// the *i*th tuple are the *i*th elements of each underlying sequence. The
+/// following example uses the `zip(_:_:)` function to iterate over an array
+/// of strings and a countable range at the same time:
```
Comment (subdan) [inline Sources/FigmaExportCore/Helpers/Zip4Sequence.swift:7]: Where did you copy the code from in this file?

**Review by subdan — CHANGES_REQUESTED** [review]
1. Fix typo `hight` → `high`
2. Fix issue I point in the comment about dark appearance.
3. Add unit tests for `XcodeColorExporter` which checks exporting high contrast colors for `useSingleFile` is `true` and `false`.
4. I tried to run FigmaExport for Example project and it fails to export with the following error: `Bad asset name...`. You must fix this issue.

File: Sources/FigmaExportCore/Helpers/Zip4Sequence.swift (hunk 10 above)
Comment (alexey1312) [reply Sources/FigmaExportCore/Helpers/Zip4Sequence.swift:7]: I indicated this in the commit comment. Duplicate here:
https://github.com/amomchilov/ZipNsequence/blob/master/Zip4Sequence.swift

**Comment by alexey1312** [comment]
1 - Fix typo hight → high ✅
2 - Fix issue I point in the comment about dark appearance. ✅
3 - Add unit tests for XcodeColorExporter which checks exporting high contrast colors for useSingleFile is true and false.
It seems that testing useSingleFile in XcodeColorExporter is not very correct, since different areas of responsibility. 
I think we need to write tests on the ColorLoader, but for this it needs to be refactored.
I checked in Playgrounds, the filter for useSingleFile works correctly. ✅
4 - I tried to run FigmaExport for Example project and it fails to export with the following error: Bad asset name.... You must fix this issue. ✅

File: Sources/FigmaExportCore/Processor/AssetsProcessor.swift [hunk 11]
```diff
@@ -90,184 +90,155 @@ public struct ImagesProcessor: AssetsProcessable {
...
-
-        // 4. descriptionMismatch
+        // FoundDuplicate
+        let lightSet: Set<AssetType> = foundDuplicate(assets: light, errors: &errors, isLightAssetSet: true)
+        let darkSet: Set<AssetType> = foundDuplicate(assets: dark, errors: &errors)
+        let lightHCSet: Set<AssetType> = foundDuplicate(assets: lightHC, errors: &errors)
+        let darkHCSet: Set<AssetType> = foundDuplicate(assets: darkHC, errors: &errors)
+        // AssetNotFoundInLightPalette
+        checkSubtracting(firstAsset: lightSet, secondAsset: darkSet, errors: &errors)
+        checkSubtracting(firstAsset: lightSet, secondAsset: lightHCSet, errors: &errors)
```
Comment (subdan) [inline Sources/FigmaExportCore/Processor/AssetsProcessor.swift:182]: Method `checkSubtracting` can generate `darkAssetsNotFoundInLightPalette` error only but you specified light and light high contrast colors.

**Review by subdan — APPROVED** [review]

**Comment by subdan** [comment]
Thanks for the PR, @alexey1312! I've published a [new version](https://github.com/RedMadRobot/figma-export/releases/tag/0.33.0) of FigmaExport with this feature.

**Comment by alexey1312** [comment]
Thanks for the review, @subdan ! 🙏
